src/test/unit_tests/jest_tests/test_versioning_concurrency.test.js

```diff
@@ -181,6 +188,277 @@ describe('test versioning concurrency', () => {
+        expect(num_of_delete_markers).toBe(0);
+        const num_of_latest_versions = (versions.objects.filter(version => version.is_latest === true)).length;
+        expect(num_of_latest_versions).toBe(initial_num_of_objects);
+    }, 60000);
+
+    it.skip('concurrent puts & list versions - version id paging', async () => {
+        const bucket = 'bucket1';
+        const upload_res_arr = [];
+        const list_res_arr = [];
+        const list_err_arr = [];
+        const upload_err_arr = [];
+        const initial_num_of_versions = 50;
+        const initial_num_of_objects = 50;
+
+        for (let i = 0; i < initial_num_of_objects; i++) {
+            const key = 'key_put' + i;
+            for (let j = 0; j < initial_num_of_versions; j++) {
+                const random_data = Buffer.from(String(crypto_random_string(7)));
+                const body = buffer_utils.buffer_to_read_stream(random_data);
+                await nsfs.upload_object({ bucket: bucket, key: key, source_stream: body }, DUMMY_OBJECT_SDK);
+            }
+        }
+        const num_of_concurrency = 20;
+        for (let i = 0; i < num_of_concurrency; i++) {
+            const key = 'key_put' + i;
+            const random_data = Buffer.from(String(crypto_random_string(7)));
+            const body = buffer_utils.buffer_to_read_stream(random_data);
+            nsfs.upload_object({ bucket: bucket, key: key, source_stream: body }, DUMMY_OBJECT_SDK)
+                .then(res => upload_res_arr.push(res.etag))
+                .catch(err => upload_err_arr.push(err));
+            nsfs.list_object_versions({ bucket: bucket }, DUMMY_OBJECT_SDK)
+                .then(res => list_res_arr.push(res))
+                .catch(err => list_err_arr.push(err));
+        }
+        await P.delay(10000);
+        expect(upload_res_arr).toHaveLength(num_of_concurrency);
+        expect(list_res_arr).toHaveLength(num_of_concurrency);
+        expect(upload_err_arr).toHaveLength(0);
+        expect(list_err_arr).toHaveLength(0);
+
+        const versions = await nsfs.list_object_versions({ bucket: bucket }, DUMMY_OBJECT_SDK);
+        expect(versions.objects).toHaveLength(1000);
+        const versions1 = await nsfs.list_object_versions({ bucket, version_id_marker: versions.next_version_id_marker }, DUMMY_OBJECT_SDK);
+        expect(versions1.objects).toHaveLength(1000);
+        const versions2 = await nsfs.list_object_versions({ bucket, version_id_marker: versions1.next_version_id_marker}, DUMMY_OBJECT_SDK);
+        expect(versions2.objects).toHaveLength(520);
+
+        const num_of_delete_markers = (versions.objects.filter(version => version.delete_marker === true)).length;
+        expect(num_of_delete_markers).toBe(0);
+        const num_of_latest_versions = (versions.objects.filter(version => version.is_latest === true)).length;
+        expect(num_of_latest_versions).toBe(initial_num_of_objects);
```

Comment:
Merge the versions*.objects into single array and then check. The first versions output has only 1k entries, it won't have all the delete markers (if any) and all the latest entries.

Reply:
Also let's consider adding something like:

```js
        const version_set = new Set(merged_versions.map(version => version.version_id));
        expect(version_set.size).toBe(2520);
```

To ensure that we do get exact number of unique versions.